lib/cusparse/generic.jl

```diff
@@ -123,63 +123,33 @@ function gather!(X::CuSparseVector, Y::CuVector, index::SparseChar)
     X
 end
 
-function mv!(transa::SparseChar, alpha::Number, A::Union{CuSparseMatrixBSR{TA},CuSparseMatrixCSR{TA}},
-             X::DenseCuVector{T}, beta::Number, Y::DenseCuVector{T}, index::SparseChar, algo::cusparseSpMVAlg_t=CUSPARSE_MV_ALG_DEFAULT) where {TA, T}
-    m,n = size(A)
-
-    if transa == 'N'
-        chkmvdims(X,n,Y,m)
-    elseif transa == 'T' || transa == 'C'
-        chkmvdims(X,m,Y,n)
-    end
-
-    descA = CuSparseMatrixDescriptor(A, index)
-    descX = CuDenseVectorDescriptor(X)
-    descY = CuDenseVectorDescriptor(Y)
-
-    # operations with 16-bit numbers always imply mixed-precision computation
-    # TODO: we should better model the supported combinations here,
-    #       and error if using an unsupported one (like with gemmEx!)
-    compute_type = if version() >= v"11.4" && T == Float16
-        Float32
-    elseif version() >= v"11.7.2" && T == ComplexF16
-        ComplexF32
-    else
-        T
-    end
-
-    function bufferSize()
-        out = Ref{Csize_t}()
-        cusparseSpMV_bufferSize(handle(), transa, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
-                                descY, compute_type, algo, out)
-        return out[]
-    end
-    with_workspace(bufferSize) do buffer
-        cusparseSpMV(handle(), transa, Ref{compute_type}(alpha), descA, descX, Ref{compute_type}(beta),
-                     descY, compute_type, algo, buffer)
-    end
-    Y
-end
-
-function mv!(transa::SparseChar, alpha::Number, A::CuSparseMatrixCSC{TA}, X::DenseCuVector{T},
+function mv!(transa::SparseChar, alpha::Number, A::Union{CuSparseMatrixCSC{TA},CuSparseMatrixCSR{TA},CuSparseMatrixCOO{TA}}, X::DenseCuVector{T},
              beta::Number, Y::DenseCuVector{T}, index::SparseChar, algo::cusparseSpMVAlg_t=CUSPARSE_MV_ALG_DEFAULT) where {TA, T}
-    ctransa = 'N'
-    if transa == 'N'
-        ctransa = 'T'
-    elseif transa == 'C' && TA <: Complex
+
+    if isa(A, CuSparseMatrixCSC) && transa == 'C' && TA <: Complex
         throw(ArgumentError("Matrix-vector multiplication with the adjoint of a CSC matrix" *
                             " is not supported. Use a CSR matrix instead."))
     end
 
-    n,m = size(A)
+    if isa(A, CuSparseMatrixCSC)
+        # cusparseSpMV doesn't support CSC format with CUSPARSE.version() < v"11.6.1"
+        # cusparseSpMV supports the CSC format with CUSPARSE.version() ≥ v"11.6.1"
+        # but it doesn't work for complex numbers when transa == 'C'
```

Comment:
Why do you do the conversion unconditionally then?

Reply:
My initial plan was to support matrix-vector and matrix-matrix products with Complex matrices `A` stored in CSC format but it's not working when the matrix is not converted.
It's the only case that we don't already support so we don't have any advantage to not do the conversion (except maybe a speedup?) but I prefer to add to not do the conversion until they debug all issues with `CSC` format upstream.
They plan to fix this issue in the next release.

We will just need to replace
```julia
if isa(A, CuSparseMatrixCSC)
```
by
```julia
if isa(A, CuSparseMatrixCSC) && CUSPARSE.version() < v"11.7.4"
```